core-bundle/src/Image/Studio/Studio.php

```diff
@@ -0,0 +1,537 @@
+     */
+    private $locale;
+
+    /**
+     * @var PictureInterface|null
+     */
+    private $picture;
+
+    /**
+     * @var MetaData|null
+     */
+    private $metaData;
+
+    public function __construct(ContainerInterface $locator)
+    {
+        $this->locator = $locator;
+    }
+
+    /**
+     * Define the resource.
+     *
+     * @param string|FilesModel $identifier can be a FilesModel, a tl_files uuid/id/path or an absolute path
+     */
+    public function from($identifier): self
+    {
+        $this->filePath = $this->getFilePath($identifier, $this->filesModel);
+
+        return $this;
+    }
+
+    /**
+     * Apply a size configuration.
+     *
+     * @param int|string|array|PictureConfiguration $size a picture size configuration or reference
+     */
+    public function setSize($size): self
+    {
+        // todo: maybe move this normalization to PictureFactory or drop?
+        $this->sizeConfiguration = StringUtil::deserialize($size);
+
+        return $this;
+    }
+
+    /**
+     * Set a meta data context. This will overwrite previously set meta data.
+     *
+     * @param ContentModel|array $context
+     */
+    public function setContext($context): self
+    {
+        $metaDataFactory = $this->locator->get('contao.image.metadata_factory');
+
+        if ($context instanceof ContentModel) {
+            $this->metaData = $metaDataFactory->createFromContentModel($context);
+
+            return $this;
+        }
+
+        if (\is_array($context)) {
+            $this->metaData = $metaDataFactory->createFromArray($context);
+
+            return $this;
+        }
+
+        throw new \InvalidArgumentException('Unsupported context type.');
+    }
+
+    /**
+     * Set meta data. This will overwrite data from a previously set context.
+     */
+    public function setMetaData(MetaData $metaData): self
+    {
+        $this->metaData = $metaData;
+
+        return $this;
+    }
+
+    /**
+     * Set a custom locale.
+     */
+    public function setLocale(string $locale): self
+    {
+        $this->locale = $locale;
+
+        return $this;
+    }
+
+    /**
+     * Create a picture with the current configuration.
+     */
+    public function getPicture(): PictureInterface
+    {
+        if (null === $this->filePath) {
+            throw new \LogicException('You need to call `from()` to set a resource before creating an image.');
+        }
+
+        /** @var PictureFactoryInterface $pictureFactory */
+        $pictureFactory = $this->locator->get('contao.image.picture_factory');
+
+        // todo: Consider using caching for other results as well
+        $this->picture = $pictureFactory->create(
+            $this->filePath, $this->sizeConfiguration
+        );
+
+        return $this->picture;
+    }
+
+    /**
+     * Get the 'sources' part of the current picture. Creates one if not existing yet.
+     */
+    public function getSources(): array
+    {
+        if (null === $this->picture) {
```

Comment:
I'd put this check into `getPicture()` itself :)

Reply:
I thought about allowing a pattern like this:
```php
$studio->from('something.jpg');
$a = $studio->setSize('_a')->getPicture();
$b = $studio->setSize('_b')->getPicture();
```
But this is probably better solved with a `dirty` flag or by either settling on caching (= allowing multiple reads) or regenerating every time.

Reply:
A dirty flag sounds good to me. Not sure if every method call affects regenerating everything so with that we could control things.